File: ios/Classes/AdInstancemanager.swift
//
//  AdInstancemanager.swift
//  flutter_custom_search_ads
//
//  Created by Desson Chan on 21/7/2022.
//

import Foundation

class AdInstanceManager {
    
    let channel: FlutterMethodChannel
    var loadedAds : [Int: CustomSearchAd] = [:]
    
    init(channel: FlutterMethodChannel) {
        self.channel = channel;
    }
    
    func loadAd(adId: Int, arguments: Dictionary<String, Any>?) {
        loadedAds[adId] = CustomSearchAd(
            manager: self,
            adId: adId,
            adUnitId: arguments?["adUnitId"] as? String ?? "",
            query: arguments?["query"] as? String ?? "",
            testAd: arguments?["testAd"] as? Bool ?? false,
            channel: arguments?["channel"] as? String ?? "",
            styleId: arguments?["styleId"] as? String ?? ""
        )
        loadedAds[adId]?.load()
    }
    
    func onAdLoaded(ad: CustomSearchAd) {
        channel.invokeMethod("onAdLoaded", arguments: ["adId": ad.adId])
    }
    
    func onAdFailedToLoad(ad: CustomSearchAd) {
        channel.invokeMethod("onAdFailedToLoad", arguments: ["adId": ad.adId])
    }
    
    func onAdHeightChanged(ad: CustomSearchAd, height: Double) {
        channel.invokeMethod("onAdHeightChanged", arguments: ["adId": ad.adId, "height": height])
    }
    
    func onAdPresent(ad: CustomSearchAd) {
        channel.invokeMethod("onAdPresent", arguments: ["adId": ad.adId])
    }
    
    func onAdImpression(ad: CustomSearchAd) {
        channel.invokeMethod("onAdImpression", arguments: ["adId": ad.adId])
    }
    
    func onAdOpen(ad: CustomSearchAd) {
        channel.invokeMethod("onAdOpen", arguments: ["adId": ad.adId])
    }
    
    func onAdClosed(ad: CustomSearchAd) {
        channel.invokeMethod("onAdClosed", arguments: ["adId": ad.adId])
    }
    
    func onAdWillDismissScreen(ad: CustomSearchAd) {
        channel.invokeMethod("onAdWillDismissScreen", arguments: ["adId": ad.adId])
    }
}

File: android/src/main/kotlin/com/example/flutter_custom_search_ads/classes/AdInstanceManager.kt
package com.example.flutter_custom_search_ads.classes

import android.app.Activity
import android.content.Context
import io.flutter.plugin.common.MethodChannel
import java.util.*

import com.example.flutter_custom_search_ads.classes.CustomSearchAd
import com.google.android.gms.ads.AdListener
import io.flutter.embedding.android.FlutterActivity

class AdInstanceManager(val channel: MethodChannel, val activity: Activity) {
    var loadedAds: MutableMap<Int, CustomSearchAd> = mutableMapOf<Int, CustomSearchAd>()

    fun loadAd(adId: Int, arguments:Map<String, Any>) {
        val ad = CustomSearchAd(this, adId,
            arguments["adUnitId"] as String,
            arguments["query"] as String,
            arguments["testAd"] as Boolean,
            arguments["channel"] as String,
            arguments["styleId"] as String
        )
        loadedAds[adId] = ad

        loadedAds[adId]?.load()
    }

    fun onAdLoad(ad: CustomSearchAd) {
        channel.invokeMethod("onAdLoaded", mapOf<String, Any>(Pair("adId", ad.adId)))
    }

    fun onFailedToLoad(ad: CustomSearchAd) {
        channel.invokeMethod("onAdFailedToLoad", mapOf<String, Any>(Pair("adId", ad.adId)))
    }

    fun onAdHeightChanged(ad: CustomSearchAd, height:Double) {
        channel.invokeMethod("onAdHeightChanged", mapOf<String, Any>(Pair("adId", ad.adId), Pair("height", height)))
    }

    fun onAdPresent(ad: CustomSearchAd) {
        channel.invokeMethod("onAdPresent", mapOf<String, Any>(Pair("adId", ad.adId)))
    }

    fun onAdImpression(ad: CustomSearchAd) {
        channel.invokeMethod("onAdImpression", mapOf<String, Any>(Pair("adId", ad.adId)))
    }

    fun onAdOpen(ad: CustomSearchAd) {
        channel.invokeMethod("onAdOpen", mapOf<String, Any>(Pair("adId", ad.adId)))
    }

    fun onAdClose(ad: CustomSearchAd) {
        channel.invokeMethod("onAdClosed", mapOf<String, Any>(Pair("adId", ad.adId)))
    }

    fun onAdWillDismissScreen(ad: CustomSearchAd) {
        channel.invokeMethod("onAdWillDismissScreen", mapOf<String, Any>(Pair("adId", ad.adId)))
    }

    fun onAdClicked(ad: CustomSearchAd) {
        channel.invokeMethod("onAdClicked", mapOf<String, Any>(Pair("adId", ad.adId)))
    }
}